src/main/java/com/rovio/ingest/util/MetadataUpdater.java

```diff
@@ -91,50 +89,61 @@ private void testDbConnection() {
     }
 
     /**
-     * Updates segments in Metadata database.
-     * Same logic as {@link org.apache.druid.indexer.SQLMetadataStorageUpdaterJobHandler#publishSegments} with
-     * additional handling for (re-)init.
+     * Updates segments in Metadata database using {@link org.apache.druid.indexer.SQLMetadataStorageUpdaterJobHandler#publishSegments},
+     * with additional handling for (re-)init.
      */
     public void publishSegments(List<DataSegment> dataSegments) {
         if (dataSegments.isEmpty()) {
             LOG.warn("No segments created, skipping metadata update.");
             return;
         }
 
-        this.sqlConnector.getDBI().withHandle(handle -> {
-            handle.getConnection().setAutoCommit(false);
-            handle.begin();
-            PreparedBatch preparedBatch = handle.prepareBatch(format(INSERT_SEGMENT_SQL, this.segmentsTable));
-            for (DataSegment segment : dataSegments) {
-                preparedBatch
-                        .bind("id", segment.getIdentifier())
-                        .bind("dataSource", segment.getDataSource())
-                        .bind("created_date", DateTimes.nowUtc().toString())
-                        .bind("start", segment.getInterval().getStart().toString())
-                        .bind("end", segment.getInterval().getEnd().toString())
-                        .bind("partitioned", !(segment.getShardSpec() instanceof NoneShardSpec))
-                        .bind("version", segment.getVersion())
-                        .bind("used", true)
-                        .bind("payload", MAPPER.writeValueAsBytes(segment))
-                        .add();
-            }
-
-            int[] execute = preparedBatch.execute();
-            if (execute.length != dataSegments.size()) {
-                throw new IllegalStateException(format("Failed to update All segments, segment=%d, updated=%d",
-                        dataSegments.size(), execute.length));
-            }
+        try {
+            RetryUtils.Task<Void> task = () -> {
+                metadataSegmentUpdaterJobHandler.publishSegments(segmentsTable, dataSegments, MAPPER);
+                return null;
+            };
+
+            RetryUtils.retry(
+                    task,
+                    this::isLockTimeoutException,
+                    0,
+                    DEFAULT_MAX_RETRIES,
+                    null,
+                    String.format("Lock wait timeout exception while publishing segments for %s", dataSource));
+        } catch (Exception e) {
+            throw new IllegalStateException("Failed to publish segments with retry", e);
```

Comment:
Is "with retry" relevant to mention here? The method contract seems to be more general. Could be something like "all retry attempts failed" or such. Should the max retries be configurable? The name "DEFAULT_" sounds like it could be sometimes different, but it seems to be just hard-coded.